Q.1 Can you explain the logic and working of the Tower of Hanoi algorithm by writing a Java program?
How does the recursion work, and how are the movements of disks between rods accomplished?

**Logic and Working:**

The Tower of Hanoi is a classic mathematical puzzle involving three rods and a number of disks of different sizes. The goal is to move all the disks from one rod to another, following these rules:

1. Only one disk can be moved at a time.
2. A larger disk cannot be placed on top of a smaller disk.

The problem can be solved recursively by breaking it down into smaller subproblems.

**Recursive Solution:**

1. **Base Case:** If there is only one disk (n = 1), move it directly from the source rod to the target rod.
2. **Recursive Case:**
   - Move n-1 disks from the source rod to the auxiliary rod using the target rod as the intermediate.
   - Move the largest disk (n-th disk) from the source rod to the target rod.
   - Move n-1 disks from the auxiliary rod to the target rod using the source rod as the intermediate.

**Java Program:**

```java
public class TowerOfHanoi {
    public static void main(String[] args) {
        int n = 3; // Number of disks
        towerOfHanoi(n, 'A', 'B', 'C');
    }

    public static void towerOfHanoi(int n, char source, char destination, char auxiliary) {
        if (n == 1) {
            System.out.println("Move disk 1 from " + source + " to " + destination);
            return;
        }

        towerOfHanoi(n - 1, source, auxiliary, destination);
        System.out.println("Move disk " + n + " from " + source + " to " + destination);
        towerOfHanoi(n - 1, auxiliary, destination, source);
    }
}
```

**Explanation:**

- The `towerOfHanoi` function takes the number of disks (n), the source rod, the destination rod, and the auxiliary rod as input.
- The base case checks if there is only one disk. If so, it moves it directly.
- In the recursive case:
    - It recursively moves n-1 disks from the source to the auxiliary rod using the destination as the intermediate.
    - It moves the largest disk (n-th disk) from the source to the destination.
    - It recursively moves n-1 disks from the auxiliary to the destination using the source as the intermediate.

**How Recursion Works:**

The `towerOfHanoi` function calls itself recursively with smaller values of n until the base case is reached. The recursive calls build up a stack of function calls, and each call returns after completing its subproblem. This allows the algorithm to efficiently solve the problem by breaking it down into smaller, manageable steps.

**Disk Movements:**

The output of the program will show the sequence of disk movements. For example, for 3 disks:

```
Move disk 1 from A to C
Move disk 2 from A to B
Move disk 1 from C to B
Move disk 3 from A to C
Move disk 1 from B to A
Move disk 2 from B to C
Move disk 1 from A to C
```

This sequence demonstrates how the disks are moved between the rods according to the rules of the Tower of Hanoi puzzle.


Q.2 Given two strings word1 and word2, return the minimum number of operations required to convert word1
to word2.

Example 1:
Input: word1 = "horse", word2 = "ros"
Output: 3
Explanation:
horse -> rorse (replace 'h' with 'r')
rorse -> rose (remove 'r')
rose -> ros (remove 'e')


**Here's a Python implementation using dynamic programming:**

```python
def minDistance(word1, word2):
    m, n = len(word1), len(word2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(m + 1):
        dp[i][0] = i
    for j in range(n + 1):
        dp[0][j] = j

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if word1[i - 1] == word2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                dp[i][j] = min(dp[i - 1][j - 1] + 1,  # Replace
                              dp[i - 1][j] + 1,      # Delete
                              dp[i][j - 1] + 1)      # Insert

    return dp[m][n]

# Example usage:
word1 = "horse"
word2 = "ros"
result = minDistance(word1, word2)
print(result)  # Output: 3
```

**Explanation:**

1. **Create a 2D dp table:** We create a dynamic programming table `dp` of size (m+1) x (n+1) to store the minimum number of operations required to convert the first i characters of word1 to the first j characters of word2.
2. **Initialize the first row and column:** We initialize the first row and column of the `dp` table with the number of operations required to insert or delete characters.
3. **Iterate through the remaining cells:**
   - If the corresponding characters in word1 and word2 are the same, the minimum number of operations is the same as the previous diagonal cell (no operation needed).
   - Otherwise, we consider the three possible operations:
     - **Replace:** dp[i-1][j-1] + 1 (replace the current character)
     - **Delete:** dp[i-1][j] + 1 (delete the current character from word1)
     - **Insert:** dp[i][j-1] + 1 (insert the current character from word2)
   - We take the minimum of these three values and store it in the current cell of the `dp` table.
4. **Return the result:** The final value in `dp[m][n]` represents the minimum number of operations required to convert word1 to word2.


Q. 3 Print the max value of the array [ 13, 1, -3, 22, 5].

In [1]:
arr = [13, 1, -3, 22, 5]
max_val = arr[0]
for num in arr[1:]:
    if num > max_val:
        max_val = num
print("Max value:", max_val)

Max value: 22


The maximum value in the array [13, 1, -3, 22, 5] is 22.

Q.4 Find the sum of the values of the array [92, 23, 15, -20, 10].

In [2]:
arr = [92, 23, 15, -20, 10]
sum_val = 0
for num in arr:
    sum_val += num
print("Sum:", sum_val)

Sum: 120


So, the sum of the values in the array is 120.

Q.5 Given a number n. Print if it is an armstrong number or not.An armstrong number is a number if the sum
of every digit in that number raised to the power of total digits in that number is equal to the number.
Example : 153 = 1^3 + 5^3 + 3^3 = 1 + 125 + 27 = 153 hence 153 is an armstrong number. (Easy)

Input1 : 153
Output1 : Yes

Input 2 : 134
Output2 : No

In [3]:
def is_armstrong(n):
    num_str = str(n)
    num_digits = len(num_str)
    sum_val = 0
    for digit in num_str:
        sum_val += int(digit) ** num_digits
    if sum_val == n:
        return "Yes"
    else:
        return "No"

# Test cases
print(is_armstrong(153))  # Output: Yes
print(is_armstrong(134))  # Output: No

Yes
No
